src/Symfony/Component/Form/ClearableErrorsInterface.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form;
+
+/**
+ * A form element whose errors can be cleared.
+ *
+ * @author Colin O'Dell <colinodell@gmail.com>
+ */
+interface ClearableErrorsInterface
+{
+    /**
+     * Removes all the errors of this form.
+     *
+     * @param bool $deep whether to remove errors from child forms as well
+     *
+     * @return $this
+     */
+    public function clearErrors(bool $deep = false);
```

Comment:
You should type hint `: self` here

Reply:
@HeahDude , we had a discussion on returning types here: https://github.com/symfony/symfony/pull/27190#discussion_r186892825
The problem will be that when using `->clearErrors()->` on actual object implementing this interface, the IDE will not show any other methods in autocomplete, because `: self` is not `: static`. So from my opinion it's much better to leave `@return $this` phpdoc.

Reply:
You can leave the `@return $this` phpdoc and add `: self` at the same time AFAK phpdoc takes precedence with type declaration for IDE auto completion, this means we would get both the correct auto completion and strictness at the language level (that an instance of the interface is returned).

Reply:
okay, than let's use `: self` 😄

Reply:
Using `: self` on an interface forces the `Form` subclass implementation to use a `: ClearableErrorsInterace` typehint:

![image](https://user-images.githubusercontent.com/202034/41508872-77f05da6-7219-11e8-812d-3262bdac8b27.png)

    Fatal error: Declaration of Symfony\Component\Form\Form::clearErrors(bool $deep = false): Symfony\Component\Form\Form must be compatible with Symfony\Component\Form\ClearableErrorsInterface::clearErrors(bool $deep = false): Symfony\Component\Form\ClearableErrorsInterface in src/Symfony/Component/Form/Form.php on line 61


I'm not sure this is what we want.

> AFAIK phpdoc takes precedence with type declaration for IDE auto completion

I could not get this working in the latest version of PhpStorm:

![image](https://user-images.githubusercontent.com/202034/41508864-4c495d7e-7219-11e8-8f8b-207bf7ac1d6b.png)

Furthermore, I could not find any other interfaces in Symfony with a `: self` return type hint.  For these reasons I think we should leave it as-is.

Reply:
The fact that we don't have any existing interfaces with a `self` return type hint is probably because we couldn't make use of them before Symfony 4. So I would still use it here and use the docblock to be more precise.